### DISCUSSION TITLE
Quality of life improvements for macroexpand-hook developers

### DISCUSSION THREAD

**Discussion by frenchy64** [clj-kondo/clj-kondo#2388] [Ideas]
To upvote this issue, give it a thumbs up. See [this list](https://github.com/clj-kondo/clj-kondo/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) for the most upvoted issues.

- [x] I have read the [Clojure etiquette](https://clojure.org/community/etiquette) and will respect it when communicating on this platform.

**Is your feature request related to a problem? Please describe.**
I maintain many macros across my projects (hundreds?) and I would like them all to be compatible with clj-kondo with minimal effort.

My solution has been to share as much macro logic between Clojure and clj-kondo as possible by factoring out macro logic to helper namespaces. I then write a build-time script to copy these to the `.clj-kondo` directory where I can use them in macroexpand hooks. This essentially restores the ability to expand the macros from clj-kondo.

[Example](https://github.com/typedclojure/typedclojure/blob/main/script/regen-kondo.sh)

I don't write hooks in the rewrite-clj style mainly because as the libraries evolve it's twice as much work to keep the expansions and linters in sync (also a library could have dozens of macros). In practice, I have not noticed the loss of precision for non-meta forms, and I have enjoyed automatic updating of clj-kondo rules once I set everything up. As long as I expand to something that's well supported by clj-kondo, error messages are good.

I think this might be a useful workflow for other busy library developers since there isn't an extra maintenance burden. If this is something we'd like to see encouraged, I have some ideas on how to improve this workflow.

**Describe the solution you'd like**
1. support `.cljc` files in hooks
   - current workaround is to rename to `.clj` when copying to the hooks directory. Luckily the reader macros still work due to a sci quirk (IIRC?).
2. add ability for hooks to require sci code from classpath
   - say we have a macro helper ns on the classpath, we currently need to copy it to `clj-kondo.exports` in order to load it from a hook ns.
   - could we add a directive in our exported `config.edn` to look for the file on the classpath?
   - or even, could a `require` in a hooks namespace look on the classpath?
3. Document which reader conditional features are activated when loading clj-kondo hooks.
   - this changes whether you're in-process or using JVM right?

**Describe alternatives you've considered**
[Script to copy helpers](https://github.com/typedclojure/typedclojure/blob/main/script/regen-kondo.sh)

It's not feasible to make a "fake" expansions or use rewrite-clj nodes due to code drift and high maintenance burden.

**Additional context**
If it's officially encouraged, it would be nice to have an example project sharing macro implementations between sci and jvm.

I can help there.

**Comment by borkdude** [comment]
1 .cljc: this is easily fixable. yes reader macros still work in SCI because the default is to just support them instead of crashing. Pushed this to master.

2 could we add a directive in our exported config.edn to look for the file on the classpath?
Currently clj-kondo has NO notion of an application classpath. It only knows what files to lint by passing `--lint $(clj -Spath)`. Clj-kondo doesn't read `deps.edn` or `project.clj` and then calculates a classpath. Loading `.clj(c)` files from the classpath, if clj-kondo had such a concept, could lead to conflicts with app/lib code that wasn't intended for clj-kondo.
However, clj-kondo does have the notion of `:config-paths` where it looks for additional `config.edn` and hook source code. 

I've been thinking of adding something to clj-kondo around classpaths though, perhaps by just starting to support `deps.edn` projects and by specifying aliases. If we had such a thing in place, then the idea could work, perhaps in combination with a (continued in 3)

3 We could add a `:clj-kondo` reader conditional as well

**Comment by frenchy64** [comment]
> 1 .cljc: this is easily fixable. yes reader macros still work in SCI because the default is to just support them instead of crashing. Pushed this to master.

Thanks!

> However, clj-kondo does have the notion of :config-paths where it looks for additional config.edn and hook source code. I've been thinking of adding something to clj-kondo around classpaths though<..>

Yes I think I can see how these are related. IIUC we could tack on an extra config.edn directive to copy hooks from places outside the export folder in the initial import, or we could do this implicitly by looking on the classpath (TBD how!) for non-exported hook namespaces.

> 3 We could add a :clj-kondo reader conditional as well

I will break the problem apart a bit.

I think there are typically 3 different modes here for each macro:
1. expand to real code in JVM (normal use)
2. expand to real code in SCI (to be used to _implement_ a hook)
3. expand to dummy code for a `:macroexpand` hook (to _be_ a macroexpand hook)

Maybe (2) is the least useful but IMO it should be supported.

Now the problem statement is: I want to use my macro in all 3 contexts but only write the macro once.

To sketch a solution, could we smuggle extra config into &env or &form?

```clojure
;; in src/
(ns my-ns
  (:require #?(:clj-kondo [clj-kondo.macroexpand-helper :refer [hook-expansion?]]))

(defn macro-helper [i-should-expand-to-fake-code?]
  ...)

(defmacro macro []
  (macro-helper #?(:clj-kondo (hook-expansion? &env)
                   :default false)))
```

Here, `hook-expansion?` returns true if the current macro is being used as a hook. We could imagine SCI associating some flag/metadata onto `&env` just before using a macro as a `:macroexpand` hook, but _not_ when loading the implementation of hooks.

I think it might be important for this flag to be lexically scoped (and not a dynamic variable) ensure any macros used to _implement_ `macro-helper` are expanded to real code (to cleanly distinguish mode 2 and 3). Hmm does SCI expand macros at runtime?

Anyway, this allows the macro to be used both to _build_ hooks and also to _be_ a hook. Something like this, for example, would mean the macro cannot be used to build hooks assuming mode (3) expands to garbage semantics:

```clojure
(defmacro macro []
  (macro-helper #?(:clj-kondo true
                   :default false)))
```

**Comment by borkdude** [comment]
> Hmm does SCI expand macros at runtime?

It expands macros in a similar way as JVM Clojure where compile-time and runtime are interleaved.
